packages/react-hooks/README.md

```diff
@@ -102,3 +102,50 @@ function Score({value}) {
   );
 }
 ```
+
+### `useToggle()`
+
+This hook will provide a boolean state value and a memoised callback to toggle it. It accepts one argument that is the initial value of the state.
+This is useful for toggling the active state of modals and popovers. If you need for force the value to be true or false regardless of the current state then consider using `useForcibleToggle()`.
+
+```tsx
+function MyComponent({value}) {
+  const [isActive, toggleIsActive] = useToggle(false);
+  const activeText = isActive ? 'true' : 'false';
+
+  return (
+    <>
+      <p>Value: {activeText}</p>
+      <button onClick={toggleIsActive}>Toggle</button>
+    </>
+  );
+}
+```
+
+### `useForcibleToggle()`
+
+This hook will provide a boolean state value and a set of memoised callbacks to toggle it, force it to true and force it to false. It accepts one argument that is the initial value of the state.
```

Comment:
I know it's not how we usually structure hook APIs, but the other functions feel so much like nice-to-haves that I wonder if it would be nicer DX to have a single hook which exposes the "bonus" callbacks as a third return value?

```tsx
const [active, toggle] = useToggle();
const [forceyActive, forceyToggle, {setTrue, setFalse}] = useToggle();
```

Reply:
Sometimes you want just the forcing functions and not the toggle though, and then things start to look a little unintuitive:

```ts
// weird
const [forceyActive, , {setTrue, setFalse}] = useToggle();
// slightly less weird, but still weird
const [forceyActive, _, {setTrue, setFalse}] = useToggle();
```

Reply:
It seems to me that the only reason to have the `forcible` option in the first place is because the `useToggle` API takes away the power that `useState` gives you. 

One solution could just be having `useToggle` be a simple hook that wraps a `useState` call and gives you the `toggle` function, and then have `useToggleState` be the hook that both creates a state *and* gives you the toggle. Then you aren't being opionated about having "forcible" callbacks, you're just giving them a maximally flexible API, and a slightly more opinionated one built on that.

### example

When you don't need regular state setting at all use `useToggleState` for convenience

```tsx
  const [shown, toggleShown] = useToggleState();

  return (
    <Button onClick={toggleShown}>toggle thing be shown</Button>
```

When you need both toggling *and* regular setting just wrap your state setter

```tsx
const [thingShown, setThingShown] = useState(false);
const toggle = useToggle(setThingShown);

return (
<>
 <Button onClick={() => setThingShown(true)}>set it open</Button>
 <Button onClick={toggle}>toggle it</Button>
```

and if you don't need toggling just like, use state

```tsx
const [modalOpen, setModalOpen] = useState(false);

return (
 <Button onClick={() => setModalOpen(true)}>set it open</Button>
```

Reply:
It's veeery subtle but  

```jsx
const [thingShown, setThingShown] = useState(false);
 <Button onClick={() => setThingShown(true)}>set it open</Button>
```

Results in a new handler being created every render, which if you pass it into a pure component will break it's pure optimisations, e.g. if Button was pure above, it would always rerender because the onClick prop differs every time. You'd need to define `const setThingShownTrue = useCallback(() => setThingShown(state => true), []);` and pass that to onClick to not break the pure check.

The utility of these two hooks is that it encapsulates those useCallback wrappings so people don't have to think about forgetting that useCallback, I think that's a key part of this.

Reply:
`useCallback` is usually premature optimization IMO so I'd rather we didn't do it for simple stuff like this by default

Reply:
been trying to discourage it in general ;)